node/circuits.js

```diff
@@ -0,0 +1,238 @@
+    if (!circuits) {
+        circuits = new EndpointCircuits(self.root);
+        self.circuitsByCallerName['$' + callerName] = circuits;
+    }
+    return circuits.getCircuit(callerName, serviceName, endpointName);
+};
+
+ServiceCircuits.prototype.getCircuitTuples = function getCircuitTuples() {
+    var self = this;
+    var circuits = Object.keys(self.circuitsByCallerName).map(function (callerName) {
+        return self.circuitsByCallerName[callerName].getCircuitTuples();
+    });
+    return Array.prototype.concat.apply([], circuits);
+};
+
+function Circuits(options) {
+    var self = this;
+    self.circuitsByServiceName = {};
+    self.stateOptions = {
+        nextHandler: alwaysShouldRequestHandler,
+        timers: options.timers,
+        random: options.random,
+        period: options.period,
+        maxErrorRate: options.maxErrorRate,
+        minRequests: options.minRequests,
+        probation: options.probation
+    };
+    self.shouldRequestOptions = {};
+    self.egressNodes = options.egressNodes;
+}
+
+Circuits.prototype.getCircuit = function getCircuit(callerName, serviceName, endpointName) {
+    var self = this;
+    var circuits = self.circuitsByServiceName['$' + serviceName];
+    if (!circuits) {
+        circuits = new ServiceCircuits(self);
+        self.circuitsByServiceName['$' + serviceName] = circuits;
+    }
+    return circuits.getCircuit(callerName, serviceName, endpointName);
+};
+
+Circuits.prototype.getCircuitTuples = function getCircuitTuples() {
+    var self = this;
+    var circuits = Object.keys(self.circuitsByServiceName).map(function (serviceName) {
+        return self.circuitsByServiceName[serviceName].getCircuitTuples();
+    });
+    return Array.prototype.concat.apply([], circuits);
+};
+
+Circuits.prototype.handleRequest = function handleRequest(req, buildRes, nextHandler) {
+    var self = this;
+    // Default the caller name.
+    // All callers that fail to specifiy a cn share a circuit for each sn:en
+    // and fail together.
+    var callerName = req.headers.cn || 'yunocn';
+    var serviceName = req.serviceName;
+    if (!serviceName) {
+        return buildRes().sendError('BadRequest', 'All requests must have a service name');
+    }
+    return req.withArg1(function withArg1() {
+        var circuit = self.getCircuit(callerName, serviceName, String(req.arg1));
+        return circuit.handleRequest(req, buildRes, nextHandler);
+    });
+};
+
+// Called upon membership change to collect services that the corresponding
+// exit node is no longer responsible for.
+Circuits.prototype.updateServices = function updateServices() {
+    var self = this;
+    var serviceNames = Object.keys(self.circuitsByServiceName);
+    for (var index = 0; index < serviceNames.length; index++) {
+        var serviceName = serviceNames[index];
+        if (!self.egressNodes.isExitFor(serviceName)) {
+            delete self.circuitsByServiceName[serviceName];
+        }
+    }
+};
+
+function Circuit() {
```

Comment:
I'd really prefer to see the constructor take at least the three primary parts of identity (the three names) and fill them in, and push down the caller name default ("yunocn" modulo shed) here.